### ISSUE TITLE
Issue with popt 1.19 and sgdisk

### ISSUE THREAD

**Issue by bryteise** [rpm-software-management/popt#80]
I'm noticing some corruption with sgdisk linking against the 1.19 popt release.

In the options parsing `[GPTDataCL::DoOptions](https://sourceforge.net/p/gptfdisk/code/ci/master/tree/gptcl.cc#l66)` the device from `poptGetArg` is a bogus string. Not really sure why as this seems to work in a more trivial example for me but I do see valgrind being unhappy when running sgdisk under it. `poptGetNextOpt` allocs and `poptResetContext` frees in the `GPTDataCL::DoOptions` function as part of a few invalid read chains. I note that if I switch to popt 1.18 the corruption goes away (as do the valgrind errors).

The [loop](https://sourceforge.net/p/gptfdisk/code/ci/master/tree/gptcl.cc#l161) doing the complex work in sgdisk could be doing something wrong and it wasn't a problem until the latest release too I suppose.

**Comment by pmatilai** [comment]
Yes - gptfdisk (and various other things) assume a value returned by poptGetArg() will remain usable forever because popt has always (?) leaked that value. The leak is fixed in 1.19 but unfortunately it breaks these applications, eg gptfdisk expects the a string from poptGetArg() to remain valid across poptResetContext(). There's a partial fix in gptfdisk but it's not in any released version yet (AFAICS).
Fedora has this downstream fix: https://src.fedoraproject.org/fork/pmatilai/rpms/gdisk/raw/71100f6fd84dd21f1346917e4fa84a2eccdb70c3/f/gdisk-1.0.9-poptmisuse.patch and i think Debian has similar too.

It's funny, we haven't actually announced popt 1.19 yet, only been preparing for it, but folks are still already finding it :sweat_smile:

**Comment by bryteise** [comment]
Ah thanks! Well people are obviously interested your new release =).

**Comment by cgzones** [comment]
Isn't it already fixed upstream: https://sourceforge.net/p/gptfdisk/code/ci/5d5e76d369a412bfb3d2cebb5fc0a7509cef878d/

**Comment by havardAasen** [comment]
@pmatilai Do you remember if there was more packages that needed to be fixed, after the release of version 1.19?

**Comment by pmatilai** [comment]
@cgzones , that upstream fix is incomplete at best: it doesn't handle `poptGetArg()` returning NULL, which can happen. It also doesn't free the now application allocated memory so it introduces a minor leak (the Fedora patch is similarly flawed in that).

**Comment by pmatilai** [comment]
@havardAasen https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=941814 mentions svox (whatever that is) at least. On Fedora authselect was flawed similarly to gdisk, but that's fixed now in their upstream release.

**Comment by elboulangero** [comment]
@havardAasen  I just bumped into that with `testparms` from Samba and reported it on the Debian bugtracker: https://bugs.debian.org/1022826

Now it seems easy to fix, however it seems that Samba is a heavy user of `poptGetArg()`:

```
$ git grep poptGetArg | wc -l
195
```

**Comment by elboulangero** [comment]
Fixed already, sorry for the noise :)

**Comment by conqp** [comment]
`libnewt` seems to be [affected, too](https://bbs.archlinux.org/viewtopic.php?id=281183).

**Comment by pmatilai** [comment]
Thanks @conqp for reporting the newt case. The flaw there is rather different from the others:
https://pagure.io/newt/blob/master/f/whiptail.c#_236 changes an allegedly const (popt internal) pointer to point at a static buffer in whiptail. And obviously, trying to free that "static char buf[20]" from poptFreeContext() does not end well.

**Comment by pmatilai** [comment]
Okay it's been about a year with no new reports, I think this ticket has served its purpose.

Hopefully we don't have to do such a thing again, breaking existing users in a minor release is terrible but, all the breakages have been due to what is quite clearly API abuse. That one can get away with a borderline thing doesn't mean it's a good idea.
